File: src/analysis.ts
import swc from '@swc/core'
import fs from 'node:fs/promises'
import { join } from 'node:path'
import { isImportDeclaration, isImportSpecifier } from './ast.js'
import { parse } from './parser.js'
import { Reporter } from './reporters/Reporter.js'
import { stringMatches } from './utils.js'
import { walk } from './walker.js'

export interface Result {
  filename: string
  source: string
  specifiers: swc.NamedImportSpecifier[]
}

export function getImports(program: swc.Module, source: string | RegExp) {
  return program.body
    .filter(isImportDeclaration)
    .filter((node) => stringMatches(node.source.value, source))
}

export function getSpecifiers(
  imports: swc.ImportDeclaration[],
  name: string | RegExp
) {
  return imports.flatMap((node) =>
    node.specifiers.filter(isImportSpecifier).filter((node) => {
      const identifier = node.imported ?? node.local
      return stringMatches(identifier.value, name)
    })
  )
}

export interface AnalyzeOptions {
  cwd: string
  name: string
  source: string
}

export async function analyze(
  reporter: Reporter,
  { cwd, name, source }: AnalyzeOptions
) {
  for (const filename of await walk({ cwd })) {
    console.log(filename)
    const sourceCode = await fs.readFile(filename, 'utf8')

    try {
      const program = await parse(filename, sourceCode)
      const imports = getImports(program, source)
      const specifiers = getSpecifiers(imports, name)

      if (!specifiers.length) continue
      reporter.progress({
        filename,
        source: sourceCode,
        specifiers,
      })
    } catch (e) {
      console.log(`Failed to parse ${filename}`)
    }
  }
}

File: src/analysis.rs
use std::path::PathBuf;

use swc_ecma_ast as ast;

use crate::logger::Logger;
use crate::parser::ParsedModule;
use crate::utils;

pub struct AnalysisRequest {
    pub path: PathBuf,
    pub source: String,
    pub specifier: Option<String>,
}

pub struct AnalysisResponse<'a> {
    pub request: &'a AnalysisRequest,
    pub parsed: &'a ParsedModule<'a>,
    pub imports: &'a Vec<&'a ast::ImportDecl>,
    pub specifiers: Option<&'a Vec<&'a ast::ImportNamedSpecifier>>,
}

fn get_imports<'a>(
    module: &'a ast::Module,
    request: &AnalysisRequest,
) -> Vec<&'a ast::ImportDecl> {
    return module
        .body
        .iter()
        .filter_map(utils::is_module_decl)
        .filter_map(utils::is_import_decl)
        .filter(|import| import.src.value == request.source)
        .collect::<Vec<_>>();
}

fn get_specifiers<'a>(
    imports: &'a Vec<&ast::ImportDecl>,
    name: &String,
) -> Vec<&'a ast::ImportNamedSpecifier> {
    return imports
        .iter()
        .flat_map(|import| import.specifiers.iter())
        .filter_map(utils::is_named_specifier)
        .filter(|specifier| specifier.local.sym.eq(name))
        .collect::<Vec<_>>();
}

pub fn analyze<T: Logger>(
    parsed: &ParsedModule,
    request: &AnalysisRequest,
    logger: &mut T,
) {
    let imports = get_imports(&parsed.module, &request);
    if imports.len() == 0 {
        return;
    }

    let specifiers = match &request.specifier {
        Some(specifier) => {
            let specifiers = get_specifiers(&imports, &specifier);

            match specifiers.len() {
                0 => return,
                _ => Some(specifiers),
            }
        }
        None => None,
    };

    logger.log(AnalysisResponse {
        imports: &imports,
        parsed,
        request,
        specifiers: specifiers.as_ref(),
    });
}
